dask/dataframe/shuffle.py

```diff
@@ -785,25 +785,18 @@ def collect(p, part, meta, barrier_token):
 
 
 def set_partitions_pre(s, divisions, ascending=True, na_position="last"):
-    try:
-        if ascending:
-            partitions = divisions.searchsorted(s, side="right") - 1
-        else:
-            partitions = len(divisions) - divisions.searchsorted(s, side="right") - 1
-    except TypeError:
-        # `searchsorted` fails if `s` contains nulls and strings
-        partitions = np.empty(len(s), dtype="int32")
-        not_null = s.notna()
-        if ascending:
-            partitions[not_null] = divisions.searchsorted(s[not_null], side="right") - 1
-        else:
-            partitions[not_null] = (
-                len(divisions) - divisions.searchsorted(s[not_null], side="right") - 1
-            )
+    partitions = np.zeros_like(s, dtype="int32")
+    not_null = s.notna().values
+    if ascending:
+        partitions[not_null] = divisions.searchsorted(s[not_null], side="right") - 1
+    else:
+        partitions[not_null] = (
+            len(divisions) - divisions.searchsorted(s[not_null], side="right") - 1
+        )
+    partitions[s.isna().values] = len(divisions) - 2 if na_position == "last" else 0
```

Comment:
I guess we only need to do this for `na_position == "last"` if partitions is created with `np.zeros`.  With that said, do we now if zeros is the best solution here for the general case? I know I originally suggested it, but I haven't run any tests locally :)

Reply:
Not sure - I can try some rough benchmarks with/without this change and with `zeros` vs. `empty` to get a sense of the performance implications